desc/geometry/curve.py

```diff
@@ -533,3 +559,217 @@ def set_coeffs(self, n, r=None):
             idx = self.r_basis.get_idx(0, 0, nn)
             if rr is not None:
                 self.r_n = put(self.r_n, idx, rr)
+
+
+class SplineXYZCurve(Curve):
+    """Curve parameterized by spline knots in X,Y,Z.
+
+    Parameters
+    ----------
+    X, Y, Z: array-like
+        points for X, Y, Z describing a closed curve
+    knots : ndarray
+        arbitrary curve parameter values to use for spline knots,
+        should be a monotonic, 1D ndarray of same length as the input X,Y,Z.
+        If None, defaults to using an equal-arclength angle as the knots
+        If supplied, will be rescaled to lie in [0,2pi]
+    method : str
+        method of interpolation
+
+        - ``'nearest'``: nearest neighbor interpolation
+        - ``'linear'``: linear interpolation
+        - ``'cubic'``: C1 cubic splines (aka local splines)
+        - ``'cubic2'``: C2 cubic splines (aka natural splines)
+        - ``'catmull-rom'``: C1 cubic centripetal "tension" splines
+        - ``'cardinal'``: C1 cubic general tension splines. If used, default tension of
+          c = 0 will be used
+        - ``'monotonic'``: C1 cubic splines that attempt to preserve monotonicity in the
+          data, and will not introduce new extrema in the interpolated points
+        - ``'monotonic-0'``: same as `'monotonic'` but with 0 first derivatives at both
+          endpoints
+
+    name : str
+        name for this curve
+
+    """
+
+    _io_attrs_ = Curve._io_attrs_ + ["_X", "_Y", "_Z", "_basis", "_transform"]
+
+    def __init__(
+        self,
+        X,
+        Y,
+        Z,
+        knots=None,
+        method="cubic",
+        name="",
+    ):
+        super().__init__(name)
+        X, Y, Z = np.atleast_1d(X), np.atleast_1d(Y), np.atleast_1d(Z)
+
+        assert np.allclose(X[-1], X[0], atol=1e-14), "Must pass in a closed curve!"
+        assert np.allclose(Y[-1], Y[0], atol=1e-14), "Must pass in a closed curve!"
+        assert np.allclose(Z[-1], Z[0], atol=1e-14), "Must pass in a closed curve!"
+        self._X = X
+        self._Y = Y
+        self._Z = Z
+
+        if knots is None:
+            # find equal arclength angle-like variable, and use that as theta
+            # L_along_curve / L = theta / 2pi
+            lengths = jnp.sqrt(
+                (X[0:-1] - X[1:]) ** 2 + (Y[0:-1] - Y[1:]) ** 2 + (Z[0:-1] - Z[1:]) ** 2
+            )
+            thetas = 2 * jnp.pi * np.cumsum(lengths) / jnp.sum(lengths)
```

Comment:
You're mixing calls to `jnp` and `np` here. just pointing out if your intention was to use only numpy. also don't need to change but i like `np.diff(X)` more than `X[0:-1] - X[1:]`.